go/packages/dev-mode/main_test.go

```diff
@@ -0,0 +1,287 @@
+	Messages  []ValidationLogMessage `json:"message"`
+}
+
+type ValidationResult struct {
+	Register  RegisterPayload        `json:"register"`
+	RequestId string                 `json:"requestId"`
+	Logs      []ValidationLogPayload `json:"logs"`
+}
+
+var port = 9001
+var region = "us-east-1"
+var functionName = "test-function"
+
+func TestMain(m *testing.M) {
+	svr := u.StartServer("test-function", "us-east-1", 9001)
+	os.Setenv("AWS_LAMBDA_RUNTIME_API", fmt.Sprintf("127.0.0.1:%d", port))
+	os.Setenv("AWS_LAMBDA_FUNCTION_VERSION", "$LATEST")
+	os.Setenv("AWS_REGION", region)
+	os.Setenv("AWS_LAMBDA_FUNCTION_NAME", functionName)
+	os.Setenv("SLS_DEBUG_EXTENSION", "1")
+	os.Setenv("SLS_TEST_EXTENSION_INTERNAL_LOG", "1")
+	os.Setenv("SLS_TEST_EXTENSION", "1")
+
+	code := m.Run()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	err := svr.Shutdown(ctx)
+	if err != nil {
+		fmt.Println(err)
+	}
+	fmt.Println("Server exiting")
+	os.Exit(code)
+}
+
+func extensionReady() {
+	u.SubmitLogs([]byte(`[{
+		"type": "platform.extension",
+		"record": {
+			"name": "dev-mode-extension",
+			"state": "Ready",
+			"events": ["INVOKE", "SHUTDOWN"]
+		}
+	}]`))
+}
+
+func extensionInvoke(requestId string) {
+	payload := fmt.Sprintf(`{
+    "eventType": "INVOKE",
+    "requestId": "%s"
+	}`, requestId)
+	u.SendPost(fmt.Sprintf("http://127.0.0.1:%d/extension/status", port), []byte(payload))
+}
+
+func extensionShutdown(requestId string) {
+	payload := fmt.Sprintf(`{
+    "eventType": "SHUTDOWN",
+    "requestId": "%s"
+	}`, requestId)
+	u.SendPost(fmt.Sprintf("http://127.0.0.1:%d/extension/status", port), []byte(payload))
+}
+
+func resetValidations() {
+	u.SendPost(fmt.Sprintf("http://127.0.0.1:%d/reset", port), []byte("{}"))
+}
+
+func getValidations() ValidationResult {
+	validationData := ValidationResult{}
+	for {
+		response, _ := http.Get(fmt.Sprintf("http://127.0.0.1:%d/validations", port))
+		jsonData, _ := ioutil.ReadAll(response.Body)
+		json.Unmarshal([]byte(jsonData), &validationData)
+		if validationData.RequestId == "" {
+			time.Sleep(2 * time.Second)
+		} else {
+			break
+		}
+	}
+	return validationData
+}
+
+func extensionPlatformStart(requestId string) {
+	payload := fmt.Sprintf(`{
+    "type": "platform.start",
+		"record": {
+			"requestId": "%s",
+			"version": "$LATEST"
+		}
+	}`, requestId)
+	u.SubmitLogs([]byte(payload))
+}
+
+func extensionPlatformRuntimeDone(requestId string) {
+	payload := fmt.Sprintf(`{
+    "type": "platform.runtimeDone",
+		"record": {
+			"requestId": "%s",
+			"status": "success"
+		}
+	}`, requestId)
+	u.SubmitLogs([]byte(payload))
+}
+
+func postLogs(logs string) {
+	u.SubmitLogs([]byte(logs))
+}
+
+func TestSuccessfulLogForwarding(t *testing.T) {
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func() {
+		ExternalExtension()
+		wg.Done()
+	}()
+
+	// First invocation
+	requestId := uuid.New().String()
+	extensionReady()
+	extensionPlatformStart(requestId)
```

Comment:
It'll be nice to have regular success case covered (where `platformStart` follows `invoke`)

Reply:
I could be misunderstanding what you are asking here but the next line is the invoke call right?

Reply:
My point is, that ideally we should have separate tests for three scenarios:
- `invoke`, `start`, `runtimeEnd`
- `start`, `invoke`, `runtimeEnd`
- `start`, `runtimeEnd`, `invoke`

Reply:
Okay I created three tests for each of these scenarios. I updated the names to the tests so the names would match basically exactly what you posted above 👍